src/org/traccar/protocol/FlespiProtocolDecoder.java

```diff
@@ -0,0 +1,116 @@
+
+    public FlespiProtocolDecoder(FlespiProtocol protocol) {
+        super(protocol);
+    }
+
+    @Override
+    protected Object decode(
+            Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+
+        HttpRequest request = (HttpRequest) msg;
+        JsonArray result = Json.createReader(new StringReader(request.getContent().toString(StandardCharsets.UTF_8)))
+                .readArray();
+        List<Position> positions = new LinkedList<>();
+        Log.debug(String.format("messages received msgs_count=%d", result.size()));
+        for (int i = 0; i < result.size(); i++) {
+            JsonObject message = result.getJsonObject(i);
+            String ident = message.getString("ident");
+            if (ident == null) {
+                continue;
+            }
+            DeviceSession deviceSession = getDeviceSession(channel, remoteAddress, ident);
+            if (deviceSession == null) {
+                continue;
+            }
+            Position position = new Position();
+            position.setDeviceId(deviceSession.getDeviceId());
+            decodePosition(message, position);
+            positions.add(position);
+        }
+
+        sendResponse(channel, HttpResponseStatus.OK);
+        return positions;
+    }
+
+    private void sendResponse(Channel channel, HttpResponseStatus status) {
+        if (channel != null) {
+            HttpResponse response = new DefaultHttpResponse(HttpVersion.HTTP_1_1, status);
+            response.headers().add(HttpHeaders.Names.CONTENT_LENGTH, 10);
+            response.setContent(ChannelBuffers.copiedBuffer("Hello namo", CharsetUtil.US_ASCII));
+            channel.write(response);
+        }
+    }
+
+    private void decodePosition(JsonObject msg, Position position) {
+        position.setProtocol("flespi");
+
+        position.setTime(new Date((long) msg.getJsonNumber("timestamp").doubleValue() * 1000));
+        JsonNumber lat = msg.getJsonNumber("position.latitude");
+        JsonNumber lon = msg.getJsonNumber("position.longitude");
+        position.setLatitude((lat != null && lon != null) ? lat.doubleValue() : 0);
+        position.setLongitude((lat != null && lon != null) ? lon.doubleValue() : 0);
+
+        JsonNumber speed = msg.getJsonNumber("position.speed");
+        position.setSpeed(speed != null ? speed.doubleValue() : 0);
+        if (position.getSpeed() == 111) {
+            position.set(Position.KEY_ALARM, Position.ALARM_GENERAL);
+        }
+
+        JsonNumber course = msg.getJsonNumber("position.direction");
+        position.setCourse(course != null ? course.doubleValue() : 0);
+
+        JsonNumber altitude = msg.getJsonNumber("position.altitude");
+        position.setAltitude(altitude != null ? altitude.doubleValue() : 0);
+
+        int satellites = msg.getInt("position.satellites", 0);
+        position.setValid(position.getLatitude() != 0 && position.getLongitude() != 0 && satellites >= 3);
```

Comment:
This doesn't look like a correct check for validity.

Reply:
Corrected: position validity is taken from "position.valid" field of message json-object.
(true by default)